File: sdk/server/ruby/lib/att/payment.rb
=begin rdoc
/**
 * @class Sencha.ServiceProvider.Att
 */
=end
module Sencha
  module ServiceProvider
    module Att

=begin rdoc
/**
 * requestChargeAuth
 *
 * @param {String} access_token The OAuth access token
 * @param {String} type Payment type
 * @param {Hash} payment_details A hash of payment options that include:
 *   @param {Number} payment_details.amount The payment amount
 *   @param {String} payment_details.product_id A product identifier
 *   @param {String} payment_details.transaction_id A unique transaction ID
 *   @param {String} payment_details.description A description of the payment
 *
 * @method requestChargeAuth
 */
=end
      def requestChargeAuth(access_token, type, payment_details)

        if payment_details[:MerchantPaymentRedirectUrl].nil?
          payment_details[:MerchantPaymentRedirectUrl] = "#{@local_server}/att/payment"
        end
        signed = signPayload(payment_details)

        if Sencha::DEBUG == :all
          @agent.log.debug "signed.data['SignedDocument'] = "
          @agent.log.debug signed.data['SignedDocument']
        end

        signed_document = signed.data['SignedDocument']
        signature = signed.data['Signature']

        paycall = chargeUser(type, signature, signed_document)
        return paycall

      end

=begin rdoc
/**
 * chargeUser
 * @param {String} type the type of charge (SUBSCRIPTION or SINGLEPAY
 * @param {String} signature
 * @param {String} signed_document the signed payload
 * @method chargeUser
 */
=end
      def chargeUser(type, signature, signed_document)
        puts 'START chargeUser'

        begin

          if type == "SUBSCRIPTION"
            type = "Subscriptions"
          end

          if type == "SINGLEPAY"
            type = "Transactions"
          end

          url = "#{@base_url}/rest/3/Commerce/Payment/#{type}?clientid=#{@client_id}&Signature=#{signature}&SignedPaymentDetail=#{signed_document}"
          chargeUrl = get_header_location(url)

          response = {
            :adviceOfChargeUrl => chargeUrl
          }

        rescue Exception => e
          Response.new e
        end
      end

=begin rdoc
/**
 * Sign a document
 * @param {Object} to_sign the JSON object to sign
 * @method signPayload
 */
=end
      def signPayload(to_sign)
        json_post("#{@base_url}/Security/Notary/Rest/1/SignedPayload?&client_id=#{@client_id}&client_secret=#{@client_secret}", to_sign.to_json)
      end

=begin rdoc
/**
 * Issues a refund for a transaction
 *
 * @param {String} access_token The OAuth access token
 * @param {String} transaction_id The id of the transaction
 * @param {String} details The json data
 *
 * @method refundTransaction
 */
=end
      def refundTransaction(access_token, transaction_id, details)
        url = "#{@base_url}/rest/3/Commerce/Payment/Transactions/#{transaction_id}?Action=refund"
        
        json_put(url, details.to_json, {"Authorization"  => "Bearer #{access_token}"})
      end

=begin rdoc
/**
 * Queries the status of a transaction
 *
 * @param {String} access_token The OAuth access token
 * @param {String} type The type of transaction (e.g. TransactionAuthCode, etc.)
 * @param {String} id
 *
 * @method transactionStatus
 */
=end
      def transactionStatus(access_token, type, id)
        url = "#{@base_url}/rest/3/Commerce/Payment/Transactions/#{type}/#{id}"
        json_get(url, {"Authorization"  => "Bearer #{access_token}"})
      end

=begin rdoc
/**
 * Queries the status of a subscription
 *
 * @param {String} access_token - the OAuth access token
 * @param {String} type - the type of transaction (e.g. TransactionAuthCode, etc.)
 * @param {String} id
 *
 * @method subscriptionStatus
 */
=end
      def subscriptionStatus(access_token, type, id)
        url = "#{@base_url}/rest/3/Commerce/Payment/Subscriptions/#{type}/#{id}"
        json_get(url, {"Authorization"  => "Bearer #{access_token}"})
      end

=begin rdoc
/**
 * Retrieves the subscription details
 *
 * @param {String} access_token The OAuth access token
 * @param {String} merchant_subscription_id The merchant subscription id
 * @param {String} consumer_id The consumer id
 *
 * @method subscriptionDetails
 */
=end
      def subscriptionDetails(access_token, merchant_subscription_id, consumer_id)
        url = "#{@base_url}/rest/3/Commerce/Payment/Subscriptions/#{merchant_subscription_id}/Detail/#{consumer_id}"
        json_get(url, {"Authorization"  => "Bearer #{access_token}"})
      end

=begin rdoc
/**
 * Retrieves a notification object
 *
 * @param {String} access_token The OAuth access token
 * @param {String} notification_id The notification id
 *
 * @method getNotification
 */
=end
      def getNotification(access_token, notification_id)
        url = "#{@base_url}/rest/3/Commerce/Payment/Notifications/#{notification_id}"
        json_get(url, {"Authorization"  => "Bearer #{access_token}"})
      end

=begin rdoc
/**
 * Stops the notification from being sent to the notification callback
 *
 *  {String} access_token The OAuth access token
 *  {String} notification_id The notification id
 *
 * @method acknowledgeNotification
 */
=end
      def acknowledgeNotification(access_token, notification_id)
        url = "#{@base_url}/rest/3/Commerce/Payment/Notifications/#{notification_id}"
        json_put(url, '', {"Authorization"  => "Bearer #{access_token}"})
      end

=begin rdoc
/**
 * Processes the payment callback redirect url
 *
 * Looks for the TransactionAuthCode and if found, sends it back to the client
 * if not found, it sends back the error info
 * @param {Object} params the params from the http response
 * @method processPaymentCallback
 */
=end
     def processPaymentCallback(params)

       response = Hash.new

       transactionAuthCode = params[:TransactionAuthCode] || params[:SubscriptionAuthCode]

       if !transactionAuthCode.nil?

         response[:success] = true
         response[:TransactionAuthCode] = transactionAuthCode

      else
        # Check for errors
        # attempt to normalize error messsages...
        response[:success] = false

         # If the payment fails to complete the app will have one set of query parameters.
         if params[:success] == "false"
           response[:error_reason] = params[:faultCode]
           response[:error_description] = params[:faultDescription]
         else
           # If the user submits a 'cancel' then we receive a different set of errors.
           response[:error] = params[:error]
           response[:error_reason] = params[:error_reason]
           response[:error_description] = params[:error_description]
         end

       end

       # send back our JSON wrapped in html
       content_type 'text/html'
       REDIRECT_HTML_PRE + response.to_json + REDIRECT_HTML_POST

     end

=begin rdoc
/**
 * Process the notification callback url
 *
 * @param {Object} request the request object from the http request
 * @param {Object} params the params object from the http request
 *
 * @method processPaymentNotification
 */
=end
     def processPaymentNotification(request, params)
       puts "Processing payment notification"
       notification_ids = Crack::XML.parse(request.body)
       notification_ids['hub:notifications']['hub:notificationId'].each do |id|
         puts "notificationId=#{id}"
         token = get_client_model_token('getNotification')
         getNotification(token, id)
         acknowledgeNotification(token, id)
       end
     end

     
     
=begin rdoc
 /**
  * Creates a JSONObject representing a Single Payment based on the given parameters
  * @param amount Amount of Transaction with a maximum of 2 decimals
  * @param category Product category
  * @param paymentDescription Short description of the entire purchase
  * @param productDescription Product name or short description
  * @param merchantTransactionId the id for the merchantTransaction.
  * @param paymentRedirectUrl url used when the transaction is completed
  * @method createSinglePaymentDescription
  */
=end
     def createSinglePaymentDescription(amount, category, paymentDescription, productDescription, merchantTransactionId, paymentRedirectUrl)
       return {
         "Amount" => amount,
         "Category" => category,
         "Channel" => "MOBILE_WEB",
         "Description" => paymentDescription,
         "MerchantTransactionId" =>  merchantTransactionId,
         "MerchantProductId"  => productDescription,
         "MerchantPaymentRedirectUrl"  => paymentRedirectUrl
       }
     end

=begin rdoc
 /**
  * Creates a JSONObject representing a Subscription Payment based on the given parameters
  * @param amount Amount of Transaction with a maximum of 2 decimals
  * @param category Product category
  * @param paymentDescription Short description of the entire purchase
  * @param productDescription Product name or short description
  * @param merchantTransactionId the id for the merchantTransaction.
  * @param paymentRedirectUrl url used when the transaction is completed
  * @param merchantSubscriptionIdList One or more subscription IDs that this transaction is associated with. Each ID has a maximum length of 12 alphanumeric characters, and commas are used to separate the values.  
  * @param freePeriods Number of free days before charging begings. Range is 0-90 
  * @method createSubscriptionPaymentDescription
  */
=end
     def createSubscriptionPaymentDescription(amount, category, paymentDescription, productDescription, merchantTransactionId, paymentRedirectUrl, merchantSubscriptionIdList, freePeriods)
       payment = createSinglePaymentDescription(amount,category,paymentDescription,productDescription, merchantTransactionId, paymentRedirectUrl)
       payment[:MerchantSubscriptionIdList] = merchantSubscriptionIdList
       payment[:SubscriptionRecurrences] = 99999
       payment[:SubscriptionPeriod] = "MONTHLY"
       payment[:SubscriptionPeriodAmount] = 1
       payment[:IsPurchaseOnNoActiveSubscription] = false
       payment[:FreePeriods] = freePeriods
       
       return payment
    end

=begin rdoc

=end    
    def refundTransactionDescription(refundReasonCode, refundReasonText)
      return {
              "TransactionOperationStatus" => "Refunded",
              "RefundReasonCode" => refundReasonCode,
              "RefundReasonText" => refundReasonText  
            }
    end

=begin rdoc

=end        
    def cancelSubscriptionDescription(cancelReasonCode, cancelReasonText)
      return {
              "TransactionOperationStatus" => "SubscriptionCancelled",
              "RefundReasonCode" => cancelReasonCode,
              "RefundReasonText" => cancelReasonText  
            }
    end
    
    end
  end
end

File: working/server/java/src/com/sencha/att/provider/Payment.java
package com.sencha.att.provider;

import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

import org.json.JSONException;
import org.json.JSONObject;

import com.sencha.att.AttConstants;

/**
 *
 * Payment class provides methods for charging the user and managing transactions and subscriptions.
 * @class com.sencha.att.provider.Payment
 *
 */
public class Payment {

    private static final String DEFAULT_PAYMENT_CHANNEL = "MOBILE_WEB";
	private static final String SUBSCRIPTIONS = "/Subscriptions/";
    private static final String TRANSACTIONS = "/Transactions/";
    private static final String NOTIFICATIONS = "/Notifications/";
    private static final String PAYMENTS_BASE_PATH = "/rest/3/Commerce/Payment";

    private static Logger log = Logger.getLogger(ServiceProviderConstants.SERVICEPROVIDERLOGGER);


    /**
     * @property {String} TYPE_SUBSCRIPTION
     */
    public static String TYPE_SUBSCRIPTION ="SUBSCRIPTION";

    /**
     * @property {String} TYPE_SINGLEPAY
     */
    public static String TYPE_SINGLEPAY ="SINGLEPAY";


    private String host;
    private String clientId;
    private String clientSecret;

    /**
     * @param {String} host
     * @param {String} clientId
     * @param {String} clientSecret
     * @constructor
     */
    public Payment(String host, String clientId, String clientSecret) {
        this.host = host;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
    }

    /**
     * First, it calls Notary to sign the payment request, then forwards the request
     * to either Subscription or Single Pay, depending on the type passed.
     * Assuming both calls are successful, a JSON object with an adviceOfChargeUrl will be returned.
     * The user will be directed to this url to authorize the charge against their account.
     *
     * @param type
     * @param paymentDetails
     * @return JSONObject with the {"adviceOfChargeUrl":https://someurl"} that the user needs to be redirected to.
     * @throws ApiRequestException
     * @method requestChargeAuth
     */
    public ApiResponse requestChargeAuth(String type, JSONObject paymentDetails) throws ApiRequestException {


        ApiResponse paycall = null;

        try {

        	//putOpt, if MerchantPaymentRedirectUrl is not already defined then add it.
            paymentDetails.putOpt("MerchantPaymentRedirectUrl", AttConstants.PAYMENTCOMPLETECALLBACK);

            NotaryProvider notary = new NotaryProvider(this.host, this.clientId, this.clientSecret);

            ApiResponse response = notary.signPayload(paymentDetails);

            String signedPaymentDetail;

            signedPaymentDetail = response.getResponse().getString("SignedDocument");
            String signature = response.getResponse().getString("Signature");

            paycall = this.chargeUser(type, signature, signedPaymentDetail);

        } catch (JSONException e) {
            e.printStackTrace();
        }



        return paycall;


    }

    /**
     * Get the advice of charge url for a signed payment request.
     *
     * @param type
     * @param signature
     * @param signedPaymentDetail
     * @return
     * @throws ApiRequestException
     * @method chargeUser
     */
    public ApiResponse chargeUser(String type, String signature, String signedPaymentDetail) throws ApiRequestException {

        String url = host + PAYMENTS_BASE_PATH;

        if(TYPE_SUBSCRIPTION.equals(type)) {
            url += "/Subscriptions";
        } else if(TYPE_SINGLEPAY.equals(type)) {
            url += "/Transactions";
        }

        url += "?clientid=" + this.clientId + "&Signature=" + signature + "&SignedPaymentDetail=" + signedPaymentDetail;



		log.info("Payment :: chargeUser GET to url: " +url);

        ApiResponse results = ApiRequestManager.get(url);

        if(results.getRedirectUrl() != null) {
            JSONObject obj = new JSONObject();
            try {
                obj.put("adviceOfChargeUrl", results.getRedirectUrl());
            } catch (JSONException e) {
                e.printStackTrace();
            }
            ApiResponse response = new ApiResponse(results.getStatusCode(), null,obj, null);
            
    		log.info("Payment :: chargeUser Response: " + response.toJson());
            return response;
        } else {
            throw new ApiRequestException("Expected a redirect from Payment call got " + results.getStatusCode(), results.getStatusCode(), results.getRawBody());
        }
    }


    /**
     * For a valid subscription id type and a valid id of that type, get the status of a subscription.
     * @param accessToken
     * @param type
     * @param subscriptionId
     * @return
     * @throws ApiRequestException
     * @method subscriptionStatus
     */
    public ApiResponse subscriptionStatus(String accessToken, String type, String subscriptionId) throws ApiRequestException {
        String url = host + PAYMENTS_BASE_PATH + SUBSCRIPTIONS + type + "/" +subscriptionId;
        Map<String, String> headers = new HashMap<String, String>();

        headers.put("Authorization", "Bearer "+accessToken);

        log.info("Payment :: subscriptionStatus GET to url: " + url);

        ApiResponse response  = ApiRequestManager.get(url, headers);
        
		log.info("Payment :: subscriptionStatus Response: " + response.toJson());

        return response;

    }

    /**
     * For a valid subscription id type and a valid id of that type, get the details of a subscription.
     *
     * @param accessToken
     * @param merchantSubscriptionId
     * @param consumerId
     * @return
     * @throws ApiRequestException
     * @method subscriptionDetails
     */
    public ApiResponse subscriptionDetails(String accessToken, String merchantSubscriptionId, String consumerId) throws ApiRequestException {
        String url = host + PAYMENTS_BASE_PATH + SUBSCRIPTIONS +merchantSubscriptionId+ "/Detail/" +consumerId;
        Map<String, String> headers = new HashMap<String, String>();

        headers.put("Authorization", "Bearer "+accessToken);
        
        log.info("Payment :: subscriptionDetails GET to url: " + url);

        ApiResponse response  = ApiRequestManager.get(url, headers);

		log.info("Payment :: subscriptionDetails Response: " + response.toJson());

        return response;

    }

    
    /**
     * Refund a transaction using a refund reason code and text
     * @param accessToken
     * @param transactionId
     * @param refundReasonCode
     * @param refundReasonText
     * @return
     * @throws JSONException
     * @throws ApiRequestException
     */
    public ApiResponse refundTransaction(String accessToken, String transactionId,String refundReasonCode, String refundReasonText) throws JSONException, ApiRequestException{
    	JSONObject refund =  new JSONObject();
    	refund.put("TransactionOperationStatus", "Refunded");
    	refund.put("RefundReasonCode", refundReasonCode);
    	refund.put("RefundReasonText", refundReasonText);
    	return refundTransaction(accessToken, transactionId, refund);
    }
    
    /**
     * Cancel a subscription using a cancel reason code and text
     * @param accessToken
     * @param transactionId
     * @param cancelReasonCode
     * @param cancelReasonText
     * @return
     * @throws JSONException
     * @throws ApiRequestException
     */
    public ApiResponse cancelSubscription(String accessToken, String transactionId, String cancelReasonCode, String cancelReasonText) throws JSONException, ApiRequestException{
    	JSONObject refund =  new JSONObject();
    	refund.put("TransactionOperationStatus", "SubscriptionCancelled");
    	refund.put("RefundReasonCode", cancelReasonCode);
    	refund.put("RefundReasonText", cancelReasonText);
    	return refundTransaction(accessToken, transactionId, refund);
    }


    /**
     * Refund a transaction id.  You must supply a JSON Object with the refund reason code.
     * @param accessToken
     * @param transactionId
     * @param details
     * @return
     * @throws ApiRequestException
     * @method refundTransaction
     */
    public ApiResponse refundTransaction(String accessToken, String transactionId, JSONObject details) throws ApiRequestException {
        String url = host + PAYMENTS_BASE_PATH + TRANSACTIONS +transactionId + "?Action=refund";

        Map<String, String> headers = new HashMap<String, String>();

        headers.put("Authorization", "Bearer "+accessToken);
        
        log.info("Payment :: refundTransaction PUT to url " + url);
        log.info("Payment :: refundTransaction Request: " + details);
        /*
         *
         * {
            "RefundReasonCode":1,
            "RefundReasonText":"Customer was not happy"
            }
         *
         */
        ApiResponse response  = ApiRequestManager.put(url, details, headers);

		log.info("Payment :: refundTransaction Response: " + response.toJson());

        
        return response;
    }


    /**
     * Get the transaction status
     *
     * @param accessToken
     * @param type  String for the type of transaction id being passed "TransactionAuthCode" or "MerchantTransactionId" or "TransactionId"
     * @param transactionId the id of the transaction to check.
     * @return
     * @throws ApiRequestException
     * @method transactionStatus
     */
    public ApiResponse transactionStatus(String accessToken, String type, String transactionId) throws ApiRequestException {
        String url = host + PAYMENTS_BASE_PATH + TRANSACTIONS + type + "/" + transactionId;

        Map<String, String> headers = new HashMap<String, String>();

        headers.put("Authorization", "Bearer "+accessToken);
        
        log.info("Payment :: transactionStatus GET to url: " + url);

        ApiResponse response  = ApiRequestManager.get(url, headers);

        log.info("Payment :: transactionStatus Response: " + response.toJson());
        
        return response;
    }


    /**
     * Gets the notification details for a notificationId
     * @param accessToken
     * @param notificationId
     * @method getNotification
     */
    public ApiResponse getNotification(String accessToken, String notificationId) throws ApiRequestException {
        String url = host + PAYMENTS_BASE_PATH + NOTIFICATIONS +notificationId;
        Map<String, String> headers = new HashMap<String, String>();

        headers.put("Authorization", "Bearer "+accessToken);
        log.info("Payment :: getNotification GET to url: " + url);

        ApiResponse response  = ApiRequestManager.get(url, headers);

		log.info("Payment :: getNotification Response: " + response.toJson());
        
        return response;
    }

    /**
     * acknowledge the receipt of a notificationId
     * @param accessToken
     * @param notificationId
     * @method acknowledgeNotification
     */
    public ApiResponse acknowledgeNotification(String accessToken, String notificationId) throws ApiRequestException {
        String url = host + PAYMENTS_BASE_PATH + NOTIFICATIONS +notificationId;
        Map<String, String> headers = new HashMap<String, String>();

        headers.put("Authorization", "Bearer "+accessToken);
        
        log.info("Payment :: acknowledgeNotification PUT to: " + url);

        ApiResponse response  = ApiRequestManager.put(url, null, headers);
        
		log.info("Payment :: acknowledgeNotification Response: " + response.toJson());

        return response;
    }

    
    
    /**
     * Creates a JSONObject representing a Subscription Payment based on the given parameters
     * @param amount Amount of Transaction with a maximum of 2 decimals
     * @param category Product category
     * @param paymentDescription Short description of the entire purchase
     * @param productDescription Product name or short description
     * @param merchantTransactionId the id for the merchantTransaction.
     * @param paymentRedirectUrl url used when the transaction is completed
     * @param merchantSubscriptionIdList One or more subscription IDs that this transaction is associated with. Each ID has a maximum length of 12 alphanumeric characters, and commas are used to separate the values.  
     * @param freePeriods Number of free days before charging begings. Range is 0-90 
     * @return
     * @throws JSONException
     */
    public static JSONObject createSubscriptionPaymentDescription(double amount, int category, String paymentDescription, String productDescription, String merchantTransactionId, String paymentRedirectUrl, String merchantSubscriptionIdList, int freePeriods) throws JSONException{
    	JSONObject payment = createSinglePaymentDescription(amount, category, paymentDescription, productDescription, merchantTransactionId, paymentRedirectUrl);
    	payment.put("MerchantSubscriptionIdList", merchantSubscriptionIdList);
    	payment.put("SubscriptionRecurrences", 99999);
    	payment.put("SubscriptionPeriod", "MONTHLY");
    	payment.put("SubscriptionPeriodAmount", 1);
    	payment.put("IsPurchaseOnNoActiveSubscription", false);
    	payment.put("FreePeriods", freePeriods);
    	return payment;
    }

    
    /**
     * Creates a JSONObject representing a Single Payment based on the given parameters
     * @param amount Amount of Transaction with a maximum of 2 decimals
     * @param category Product category
     * @param paymentDescription Short description of the entire purchase
     * @param productDescription Product name or short description
     * @param merchantTransactionId the id for the merchantTransaction.
     * @param paymentRedirectUrl url used when the transaction is completed
     * @return
     * @throws JSONException
     */
    public static JSONObject createSinglePaymentDescription(double amount, int category, String paymentDescription, String productDescription, String merchantTransactionId, String paymentRedirectUrl) throws JSONException{
    	JSONObject payment = new JSONObject();
		payment.put("Amount", amount);
		payment.put("Category", category);
		payment.put("Channel", DEFAULT_PAYMENT_CHANNEL);
		payment.put("Description", paymentDescription);
		payment.put("MerchantTransactionId", merchantTransactionId);
		payment.put("MerchantProductId", productDescription);
    	payment.put("MerchantPaymentRedirectUrl", paymentRedirectUrl);
    	return payment;
    }
    
    /**
     * Create a description JSONObject to be used on refundTransaction
     * @param refundReasonCode
     * @param refundReasonText
     * @return
     * @throws JSONException
     */
    public static JSONObject refundTransactionDescription(String refundReasonCode, String refundReasonText) throws JSONException{
    	JSONObject refund =  new JSONObject();
    	refund.put("TransactionOperationStatus", "Refunded");
    	refund.put("RefundReasonCode", refundReasonCode);
    	refund.put("RefundReasonText", refundReasonText);
    	return refund;
    }
    
    /**
     * Create a description JSONObject to be used on refundTransaction
     * @param cancelReasonCode
     * @param cancelReasonText
     * @return
     * @throws JSONException
     */
    public static JSONObject cancelSubscriptionDescription(String cancelReasonCode, String cancelReasonText) throws JSONException{
    	JSONObject refund =  new JSONObject();
    	refund.put("TransactionOperationStatus", "SubscriptionCancelled");
    	refund.put("RefundReasonCode", cancelReasonCode);
    	refund.put("RefundReasonText", cancelReasonText);
    	return refund;    	
    }

    
}
